form-flow-valtimo/src/main/kotlin/com/ritense/valtimo/formflow/web/rest/FormFlowResource.kt

```diff
@@ -40,41 +43,35 @@ class FormFlowResource(
 
         val stepInstance = instance.getCurrentStep()
 
-        return ResponseEntity.ok(
-            GetFormFlowStateResult(
-                instance.id.id,
-                FormFlowStepResult(
-                    stepInstance.id.id,
-                    stepInstance.definition.type.name,
-                    stepInstance.definition.type.properties
-                    // TODO Include the prefilled form when this functionality is available
-                )
-            )
-        )
+        return ResponseEntity.ok(GetFormFlowStateResult(instance.id.id, getStepResult(stepInstance)))
     }
 
-    @PostMapping("/{formFlowId}/step/{stepInstanceId}")
+    @PostMapping("/{formFlowId}/step/{stepInstanceId}", produces = [MediaType.APPLICATION_JSON_VALUE])
+    @Transactional
     fun completeStep(
         @PathVariable(name = "formFlowId") formFlowId: String,
         @PathVariable(name = "stepInstanceId") stepInstanceId: String,
         @RequestBody submissionData: JsonNode?
-    ): ResponseEntity<FormFlow> {
-        return ResponseEntity.ok(getStepDto())
+    ): ResponseEntity<CompleteStepResult> {
+        val instance = formFlowService.getByInstanceIdIfExists(
+            FormFlowInstanceId.existingId(UUID.fromString(formFlowId))
+        )!!
+
+        val stepInstance = instance.complete(
+            FormFlowStepInstanceId.existingId(UUID.fromString(stepInstanceId)),
+            JSONObject((submissionData ?: JsonNodeFactory.instance.objectNode()).toString())
```

Comment:
This works for what we want to do. i just wonder what we could do to prevent having to parse the same JSON twice with different libraries.

Reply:
Changed it